src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController.cs

```diff
@@ -387,21 +380,27 @@ private void ProcessItemSelectionSynchronously(NavigationBarItem item, Cancellat
                 var document = _subjectBuffer.CurrentSnapshot.AsText().GetDocumentWithFrozenPartialSemantics(cancellationToken);
                 if (document != null)
                 {
-                    var languageService = document.GetRequiredLanguageService<INavigationBarItemService>();
+                    var navBarService = document.GetRequiredLanguageService<INavigationBarItemService>();
+                    var snapshot = _subjectBuffer.CurrentSnapshot;
+                    item.Spans = item.TrackingSpans.SelectAsArray(ts => ts.GetSpan(snapshot).Span.ToTextSpan());
+                    var view = _presenter.TryGetCurrentView();
 
-                    NavigateToItem(item, document, _subjectBuffer.CurrentSnapshot, languageService, cancellationToken);
+                    if (navBarService is INavigationBarItemService2 navBarService2)
+                    {
+                        // ConfigureAwait(true) as we have to come back to UI thread in order to kick of the refresh task below.
+                        await navBarService2.NavigateToItemAsync(document, item, view, cancellationToken).ConfigureAwait(true);
+                    }
+                    else
+                    {
+                        navBarService.NavigateToItem(document, item, view, cancellationToken);
+                    }
                 }
             }
 
-            // Now that the edit has been done, refresh to make sure everything is up-to-date. At
-            // this point, we now use CancellationToken.None to ensure we're properly refreshed.
-            UpdateDropDownsSynchronously(CancellationToken.None);
-        }
-
-        private void NavigateToItem(NavigationBarItem item, Document document, ITextSnapshot snapshot, INavigationBarItemService languageService, CancellationToken cancellationToken)
-        {
-            item.Spans = item.TrackingSpans.Select(ts => ts.GetSpan(snapshot).Span.ToTextSpan()).ToList();
-            languageService.NavigateToItem(document, item, _presenter.TryGetCurrentView(), cancellationToken);
+            // Now that the edit has been done, refresh to make sure everything is up-to-date.
+            // Have to make sure we come back to the main thread for this.
```

Comment:
```suggestion
            // Have to make sure we're still on the main thread for this.
```
comment was a little confusing - we shouldn't have left the UI thread at all right?

Reply:
true.  that comment is better.